enhancements/network/bgp-ovn-kubernetes.md

```diff
@@ -0,0 +1,701 @@
+will need the ability to:
+
+* Define BGP peers, ASN, authentication, along with other common BGP configurations like communities, route filtering, etc.
+* Configure BFD
+
+FRR-K8S will serve as the main API for user configuration of BGP via FRR-K8S CRs. OVN-Kubernetes will watch for these
+CRs, and once created, OVN-Kubernetes will start monitoring the kernel routing table via netlink for routes installed
+by FRR. When FRR is using Zebra, the routes can be identified via a proto "bgp" label:
+
+```
+123.123.123.0/24 nhid 47 via 172.18.0.5 dev breth0 proto bgp metric 20
+```
+
+When these routes are seen on the node, ovnkube-controller will update the routes in NBDB for any GW routers on the
+node, filtering the route for the GW router based on if the interface on the route matches the interface in OVS. This
+will happen irrespective of gateway mode, as some features in local gateway mode rely on the GRs for sending traffic
+(Egress IP, MEG). Note, this configuration will only happen on nodes selected by the FRR-K8S CRs.
+
+#### API Changes
+
+##### OpenShift API
+
+The OpenShift API will be modified in order to allow CNO to deploy FRR and FRR-K8S CRDs:
+
+```golang
+// NetworkSpec is the top-level network configuration object.
+type NetworkSpec struct {
+    // ...
+    // deployFRR specifies whether or not the Free Range Routing (FRR) stack
+    // along with FRR-K8S API should be deployed by the operator.
+    // FRR is required for enabling for using any OpenShift features that require dynamic
+    // routing. This includes BGP support for OVN-Kubernetes and MetalLB.
+    // +optional
+    DeployFRR *bool `json:"deployFRR,omitempty"`
+}
+```
+
+##### Route Advertisements
+
+When OVN-Kubernetes detects that a FRR-K8S CR has been created for BGP peering, OVN-Kubernetes will by default
+advertise the pod subnet for each node, via creating an additive, per node FRR-K8S CR that is managed by OVN-Kubernetes.
+OVN-Kubernetes CRD:
+
+```yaml
+apiVersion: k8s.ovn.org/v1
+kind: RouteAdvertisements
+metadata:
+  name: default
+spec:
+    # networkSelector:
+    # nodeSelector:
+    # frrConfigurationSelector:
+    targetVRF: default
+    advertisements:
+      podNetwork: true
+      egressIP: true
+```      
+
+In the above example, an optional networkSelector may also be optionally supplied which will match namespaces/networks.
+When omitted, RouteAdvertisements will be applied to the default cluster network. If a selector is used that selects the
+default cluster network, it will be enabled for any namespace using the cluster default network. Additionally,
+advertisements may be limited to specific nodes via the nodeSelector.
+
+A networkSelector may select more than one network, including user-defined networks. In such a case the network subnets
+will be checked by OVN-Kubernetes to determine if there is any overlap of the IP subnets. If so, an error status will be
+reported to the CRD and no BGP configuration will be done by OVN-Kubernetes.
```

Comment:
Are the subnets in a network immutable after creation time? If that's not enforced by validation then you also need to recheck the RouteAdvertisements when the Network changes

Reply:
Changing them after creation (and after having pods assigned a cidr) can be messy I think? What is gonna happen to a pod with an IP that no longer belongs to the subnet? 

I think it's reasonable to have the subnets immutable and enforce that, possibly with [cel transition rules](https://kubernetes.io/blog/2022/09/29/enforce-immutability-using-cel/), but I'd wait @trozet

Reply:
should not be possible to change the subnet after creation time. Thanks for the comment, I will add that to the UDN CRD PR: https://github.com/openshift/enhancements/pull/1638